### DISCUSSION TITLE
Do we have input filter feature？

### DISCUSSION THREAD

**Discussion by ChuLiqiang** [resiprocate/resiprocate#296] [Q&A] [answered]
Do we have input filter feature? I will add a note header in sip header. I will check all message node header to determine whether the authentication is pass。I want not to add this feature in a lot of 200 ok response msg handle and a lot of cxx class.(I want to confirm if have a feature like Outbound Decorators in input?)

**Comment by sgodin** [comment, marked as answer]
Sorry for the slow response.  DUM has a concept of Incoming and Outbound Features to process messages for authentication.  These are used for DUM ServerAuthManager.  Unfortunately there is very little documentation of these.  See DialogUsageManager addIncomingFeature and addOutgoingFeature.

**Reply by ChuLiqiang** [reply]
I find a example in esip/dum/test/treg.cxx. I add a DumFeature in addIncomingFeature. when we recive a notify message it can not callback to virtual ProcessingResult process(Message* msg) function. But it is callback for ACK message in INVITE dialog.  addIncomingFeature is only work for dialog message? is it does`t work for pager meesage？
![image](https://user-images.githubusercontent.com/2747722/215374870-7dfe389e-373a-442f-9abd-a56673d61c14.png)



IP (tos 0x0, ttl 64, id 27764, offset 0, flags [DF], proto UDP (17), length 665)
    172.20.32.135.5161 > 172.20.32.111.sip-tls: [udp sum ok] UDP, length 637
E...lt@.@.2... ... o.)......MESSAGE sip:71000000002000000111@172.20.32.111:5061 SIP/2.0
Via: SIP/2.0/UDP 172.20.32.135:5161;branch=z9hG4bK-524287-1---0a4e955a0bc00b0d;rport
Max-Forwards: 70
To: <sip:71000000002000000111@172.20.32.111:5061>
From: <sip:71000000002000000135@7100000000>;tag=af010872
Call-ID: g4Y8EYbYubEUMgmCR71veQ..
CSeq: 2 MESSAGE
Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, NOTIFY, SUBSCRIBE, INFO, MESSAGE
Content-Type: Application/MANSCDP+xml
User-Agent: SAILING
Content-Length: 152

<?xml version="1.0"?>
<Notify>
<CmdType>Keepalive</CmdType>
<SN>224</SN>
<DeviceID>71000000002000000135</DeviceID>
<Status>OK</Status>
</Notify>

**Reply by ChuLiqiang** [reply]
I find notify is not callback. other message is ok?is it  noly work for response message？

**Reply by ChuLiqiang** [reply]
I find if message is request. it will be beak in IncomingFeature.Why request for example(Notify) can not handle by IncomingFeature.
![image](https://user-images.githubusercontent.com/2747722/215412935-34348021-1a50-41af-a0ef-4fc1976bb13f.png)

**Reply by sgodin** [reply]
Hmm the code you pasted above does not look like the code on Git main.  I don't see code on main that blocks requests.

**Reply by ChuLiqiang** [reply]
sorry we add this code by myself....
